python/cudf/cudf/__init__.py

```diff
@@ -80,35 +80,17 @@
     read_text,
 )
 from cudf.options import describe_option, get_option, set_option
+from cudf.utils._setup_numba import _setup_numba
 from cudf.utils.dtypes import _NA_REP
 from cudf.utils.utils import clear_cache, set_allocator
 
-try:
-    from cubinlinker.patch import patch_numba_linker_if_needed
-except ImportError:
-    pass
-else:
-    # Patch Numba to support CUDA enhanced compatibility.
-    # cuDF requires a stronger set of conditions than what is
-    # checked by patch_numba_linker_if_needed due to the PTX
-    # files needed for JIT Groupby Apply and string UDFs
-    from cudf.core.udf.utils import _PTX_FILE, _setup_numba_linker
+_setup_numba()
```

Comment:
Any magic in the import order is usually good to explicitly describe for readers.

```suggestion
# Must be called before importing numba.cuda.
_setup_numba()
```

Is it dangerous / invalid if a user imports `numba.cuda` in their own code before importing cudf?
- If so, will users clearly understand the problem (i.e. is a helpful error raised)?
- If not, why is this special import ordering needed? (Could it be written above, with the other `numba` import?)

Reply:
My interpretation of @brandon-b-miller's response [here](https://github.com/rapidsai/cudf/pull/13337#discussion_r1197041433) was that in fact there is no longer an explicit ordering issue because we are using the numba config API rather than an environment variable. If there is still an ordering issue then I remain confused about my question in the other thread because it seems like we would be breaking the ordering there.

Reply:
So this is still an item of concern because we've pivoted back to putting numba 0.57 into this PR. Currently the call to `_setup_numba` is placed very early on in cudf's `__init__` even before `validate_setup()` is called, as `validate_setup()` calls `rmm._gpu.runtimeGetVersion()` which actually imports `numba.cuda` as a side effect.

The order of imports _does_ matter here. In addition there's @bdice 's observation that a user importing `numba.cuda` somewhere in their workflow _before_ `cudf` is imported is likely to break cuDF. So certainly this is an issue we need to come to some consensus about before merging. 

One option I can think of that could work is detecting this case dynamically somehow and issuing a warning to the user that instructs them to set `CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY=1` in their environment. This should set the variable globally before the user's import.